src/orange/detail/BIHPartitioner.cc

```diff
@@ -0,0 +1,175 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHPartitioner.cc
+//---------------------------------------------------------------------------//
+#include "BIHPartitioner.hh"
+
+#include "BoundingBoxUtils.hh"
+
+namespace
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sort and uniquify a vector.
+ */
+void sort_and_uniquify(std::vector<celeritas::real_type>& vec,
+                       celeritas::real_type tol)
+{
+    std::sort(vec.begin(), vec.end());
+
+    auto last = std::unique(vec.begin(),
+                            vec.end(),
+                            [&](celeritas::real_type a, celeritas::real_type b) {
+                                return std::abs(a - b) < tol;
+                            });
+
+    vec.erase(last, vec.end());
+}
+}  // namespace
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from vector of bounding boxes and respective centers.
+ */
+BIHPartitioner::BIHPartitioner(VecBBox* bboxes, VecReal3* centers)
+    : bboxes_(bboxes), centers_(centers)
+{
+    CELER_EXPECT(!bboxes_->empty());
+    CELER_EXPECT(bboxes_->size() == centers_->size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine if a set of bounding boxes can be partitioned
+ */
+bool BIHPartitioner::is_partitionable(VecIndices const& indices) const
+{
+    CELER_EXPECT(!indices.empty());
+
+    auto centers = this->axes_centers(indices);
+    return std::any_of(centers.begin(),
+                       centers.end(),
+                       [](std::vector<real_type> const& centers) {
+                           return centers.size() > 1;
+                       });
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find a suitable partition for the given bounding boxes.
+ */
+BIHPartitioner::Partition
+BIHPartitioner::operator()(VecIndices const& indices) const
+{
+    CELER_EXPECT(!indices.empty());
+
+    Partition best_partition;
+    real_type best_cost = std::numeric_limits<real_type>::infinity();
+
+    auto axes_centers = this->axes_centers(indices);
```

Comment:
This is how it was originally. I think I prefer the "is_partitionable" route with short-circuiting, but if you feel strongly otherwise I can go with the null partition route.

Reply:
I think if you're making partitioner function-like and it reduces the overall work and code complexity, it's definitely the way to go.